sound/soc/sof/sof-audio.c

```diff
@@ -347,3 +467,180 @@ struct snd_sof_dai *snd_sof_find_dai(struct snd_soc_component *scomp,
 
 	return NULL;
 }
+
+static int sof_destroy_pipelines(struct device *dev)
+{
+	struct sof_audio_dev *sof_audio = sof_get_client_data(dev);
+	struct snd_sof_widget *swidget;
+	struct sof_ipc_free ipc_free;
+	struct sof_ipc_reply reply;
+	int ret = 0;
+
+	list_for_each_entry_reverse(swidget, &sof_audio->widget_list, list) {
+		memset(&ipc_free, 0, sizeof(ipc_free));
+
+		/* skip if there is no private data */
+		if (!swidget->private)
+			continue;
+
+		/* configure ipc free message */
+		ipc_free.hdr.size = sizeof(ipc_free);
+		ipc_free.hdr.cmd = SOF_IPC_GLB_TPLG_MSG;
+		ipc_free.id = swidget->comp_id;
+
+		switch (swidget->id) {
+		case snd_soc_dapm_scheduler:
+			ipc_free.hdr.cmd |= SOF_IPC_TPLG_PIPE_FREE;
+			break;
+		case snd_soc_dapm_buffer:
+			ipc_free.hdr.cmd |= SOF_IPC_TPLG_BUFFER_FREE;
+			break;
+		default:
+			ipc_free.hdr.cmd |= SOF_IPC_TPLG_COMP_FREE;
+			break;
+		}
+		ret = sof_client_tx_message(dev, ipc_free.hdr.cmd,
+					    &ipc_free, sizeof(ipc_free), &reply,
+					    sizeof(reply));
+		if (ret < 0) {
+			dev_err(dev,
+				"error: failed to free widget type %d with ID: %d\n",
+				swidget->widget->id, swidget->comp_id);
+
+			return ret;
+		}
+	}
+
+	return ret;
+}
+
+static int sof_audio_resume(struct device *dev)
+{
+	return sof_restore_pipelines(dev);
+}
+
+static int sof_audio_suspend(struct device *dev)
+{
+	return sof_set_hw_params_upon_resume(dev);
+}
+
+static int sof_audio_runtime_suspend(struct device *dev)
+{
+	return sof_destroy_pipelines(dev);
+}
+
+static const struct dev_pm_ops sof_audio_pm = {
+	SET_SYSTEM_SLEEP_PM_OPS(sof_audio_suspend, sof_audio_resume)
+	SET_RUNTIME_PM_OPS(sof_audio_runtime_suspend, sof_audio_resume, NULL)
+};
+
+static int sof_audio_probe(struct platform_device *pdev)
```

Comment:
we could use a wrapper so that we can reuse the same code as before?

e.g. 
sof_audio_probe_pdev(struct platform_device *pdev) 
{
    return sof_audio_probe(&pdev->dev)
}

that would help test without necessarily create this platform driver